### DISCUSSION TITLE
Memory leak introduced in v3.2.0

### DISCUSSION THREAD

**Discussion by calebdw** [livewire/livewire#7344] [7. Bugs (archived)] [answered]
### Livewire version

3.2.0

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

1. Install v3.2.0
2. Run pest tests

### Additional information

Hello!

It seems that there was a memory leak introduced in v3.2.0---my tests fail (very early on, not even 1/10th through the test suite) with the below error after upgrading from v3.1.0:

![image](https://github.com/livewire/livewire/assets/4176520/7a428dc1-64a0-4025-a03d-9af7af13698e)

This is affecting the entire test suite and is not specific to Livewire tests

Note that the entire test suite passes just fine on v3.1.0 so increasing the memory limit is not the solution.

Thanks!

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by calebdw** [comment]
Culprit PR: https://github.com/livewire/livewire/pull/7181

**Comment by bpotmalnik** [comment]
@calebdw Hey, I encountered this issue as well. 

After small investigation I think there is a problem on how we are setting callback on the middleware in `HandleRequests.php`. 

So after the change in [this PR](https://github.com/livewire/livewire/pull/7181) we are now resolving mechanisms out of the container. 

On of the mechanisms [HandleRequests.php](https://github.com/livewire/livewire/blob/main/src/Mechanisms/HandleRequests/HandleRequests.php) is registering some skipping functionality middleware.
https://github.com/livewire/livewire/blob/95083c0b61c1910ad28c4628e0dfa9c227eb1c41/src/Mechanisms/HandleRequests/HandleRequests.php#L24

And this skip functionality middleware is registering callback on when to skip this middleware.
https://github.com/livewire/livewire/blob/95083c0b61c1910ad28c4628e0dfa9c227eb1c41/src/Mechanisms/HandleRequests/HandleRequests.php#L36-L42

And skip middleware is registering it statically in the [middleware](https://github.com/laravel/framework/blob/f0ac18b9b4323e80fb45f81bd47747445e5e885b/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php#L54) so what I think ends up happening it we are just reregistering this callback without checking if it was previously registered, which causes and memory leak.

@calebdw Could confirm if you comment out [`skipRequestPayloadTamperingMiddleware`](https://github.com/livewire/livewire/blob/main/src/Mechanisms/HandleRequests/HandleRequests.php#L24) that memory leak is resolved?

Edit: 
Manage to replicate it with default jetstream app.

Here is the repository: https://github.com/bpotmalnik/livewire-memory-leak-demo
I've added [simple test](https://github.com/bpotmalnik/livewire-memory-leak-demo/blob/main/tests/Unit/LivewireMemoryLeakTest.php) that hits`/` page 2000 times (or more depending on how much memory is your php instance allocated). On my machine it fails around 710 repeat. 

If I comment out `skipRequestPayloadTamperingMiddleware` in my vendor folder it reaches 2000 without a problem.

**Reply by calebdw** [reply]
@bpotmalnik, that seems to have fixed it!

**Reply by calebdw** [reply]
@bpotmalnik, although I'm not sure how using the container to resolve the class causes memory to skyrocket whereas just `new`ing up the object does not? The `skipRequestPayloadTamperingMiddleware` method was still being called...

**Comment by tzurbaev** [comment]
Just encountered this bug after installing `laravel/pulse` (I was not using Livewire before). The `HandleRequests::skipRequestPayloadTamperingMiddleware` method registers middleware callbacks on each request, which is not bad by itself, but it slows down unit tests a lot.

I think it would be nice to have some static flag check inside `skipRequestPayloadTamperingMiddleware` method:

```php
function skipRequestPayloadTamperingMiddleware()
{
    if (self::$skipped) {
        return;
    }

    \Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull::skipWhen(function () {
        return $this->isLivewireRequest();
    });

    \Illuminate\Foundation\Http\Middleware\TrimStrings::skipWhen(function () {
        return $this->isLivewireRequest();
    });

    self::$skipped = true;
}
```

Update: if you're **only** using `livewire/livewire` as a dependency to `laravel/pulse` (meaning you don't have any Livewire views), you can use this workaround until this bug is fixed:

1. Add `"livewire/livewire"` to the Laravel's `dont-discover` array (in composer.json);
2. Register `LivewireServiceProvider` in your `AppServiceProvider` manually:

```php

public function register(): void
{
    if (app()->environment() !== 'testing') {
        $this->app->register(LivewireServiceProvider::class);
    }
}
```

This should not break regular requests to `/pulse` dashboard and won't introduce memory leaks while running tests. Also you might want to disable Laravel Pulse in your tests just in case (`<env name="PULSE_ENABLED" value="false"/>`).

**Comment by WayneHarris** [comment, marked as answer]
With this latest Laravel release, the issue is resolved. https://github.com/laravel/framework/releases/tag/v10.37.3
